GVRf/Framework/framework/src/main/jni/engine/renderer/renderer.cpp

```diff
@@ -121,7 +130,212 @@ void Renderer::state_sort() {
         }
     }
 }
+struct comparator {
+    std::string renderdata_properties;
+    Material* mat;
+    Material::ShaderType shader_type;
+    bool mesh_dynamic;
+};
+
+/*
+ * batch_set stores all the batches, batch_map stores the indices of batches and the respective map
+ * to get constant look up time
+ */
+std::vector<Batch*> batch_set;
+std::unordered_map<Batch*, int> batch_map;
+#define MAX_INDICES 400
+/*
+ * batch indices stores the indices in render_vector where batches are split
+ */
+std::vector<int> batch_indices;
+void createBatch(int start, int end) {
+    Batch* existing_batch = nullptr;
+
+    int size = BATCH_SIZE;
+    // get batch with least no of meshes in it
+    for (int i = start; i <= end; ++i) {
+        if (render_data_vector[i]->getBatch() != nullptr) {
+            if (render_data_vector[i]->getBatch()->getNumberOfMeshes()
+                    <= size) {
+                size = render_data_vector[i]->getBatch()->getNumberOfMeshes();
+                existing_batch = render_data_vector[i]->getBatch();
+            }
+        }
+    }
+
+    for (int i = start; i <= end; ++i) {
+        RenderData* render_data = render_data_vector[i];
+        Batch* current_batch = render_data->getBatch();
+        if (!current_batch) {
+            if (!render_data->mesh()->isDynamic()) { // mesh is static
+                // existing batch is not full
+                if (existing_batch
+                        && existing_batch->getNumberOfMeshes() < BATCH_SIZE) {
+                    // add failed because mesh is large try with next batch
+                    if (!existing_batch->add(render_data)) {
+                        Batch* new_batch = new Batch(MAX_INDICES, MAX_INDICES);
+                        new_batch->add(render_data);
+                        render_data->setBatch(new_batch);
+                        batch_set.push_back(new_batch);
+                        batch_map[new_batch] = batch_set.size() - 1;
+                        existing_batch = new_batch;
```

Comment:
Same Batch creation code in 6 locations -- for easier maintenance move this to a function.